File: javax/src/main/java/me/panpf/javax/util/Base64x.java
package me.panpf.javax.util;

import org.jetbrains.annotations.NotNull;

import java.nio.ByteBuffer;
import java.nio.charset.Charset;

/**
 * Base64 tool method
 */
@SuppressWarnings("WeakerAccess")
public class Base64x {

    @NotNull
    public static String encodeToString(@NotNull byte[] src) {
        return Base64.getEncoder().encodeToString(src);
    }

    @NotNull
    public static String encodeToString(@NotNull String text) {
        return Base64.getEncoder().encodeToString(text.getBytes(Charset.forName("UTF-8")));
    }

    @NotNull
    public static String encodeToString(@NotNull ByteBuffer byteBuffer) {
        return Base64.getEncoder().encodeToString(byteBuffer.array());
    }


    @NotNull
    public static byte[] encodeToBytes(@NotNull byte[] src) {
        return Base64.getEncoder().encode(src);
    }

    @NotNull
    public static byte[] encodeToBytes(@NotNull String text) {
        return Base64.getEncoder().encode(text.getBytes(Charset.forName("UTF-8")));
    }

    @NotNull
    public static byte[] encodeToBytes(@NotNull ByteBuffer byteBuffer) {
        return Base64.getEncoder().encode(byteBuffer.array());
    }


    public static int encodeTo(@NotNull byte[] src, @NotNull byte[] dst) {
        return Base64.getEncoder().encode(src, dst);
    }

    public static int encodeTo(@NotNull String text, @NotNull byte[] dst) {
        return Base64.getEncoder().encode(text.getBytes(Charset.forName("UTF-8")), dst);
    }

    public static int encodeTo(@NotNull ByteBuffer byteBuffer, @NotNull byte[] dst) {
        return Base64.getEncoder().encode(byteBuffer.array(), dst);
    }


    @NotNull
    public static ByteBuffer encodeToBuffer(@NotNull byte[] src) {
        return Base64.getEncoder().encode(ByteBuffer.wrap(src));
    }

    @NotNull
    public static ByteBuffer encodeToBuffer(@NotNull String text) {
        return Base64.getEncoder().encode(ByteBuffer.wrap(text.getBytes(Charset.forName("UTF-8"))));
    }

    @NotNull
    public static ByteBuffer encodeToBuffer(@NotNull ByteBuffer byteBuffer) {
        return Base64.getEncoder().encode(byteBuffer);
    }


    @NotNull
    public static String urlEncodeToString(@NotNull byte[] src) {
        return Base64.getUrlEncoder().encodeToString(src);
    }

    @NotNull
    public static String urlEncodeToString(@NotNull String text) {
        return Base64.getUrlEncoder().encodeToString(text.getBytes(Charset.forName("UTF-8")));
    }

    @NotNull
    public static String urlEncodeToString(@NotNull ByteBuffer byteBuffer) {
        return Base64.getUrlEncoder().encodeToString(byteBuffer.array());
    }


    @NotNull
    public static byte[] urlEncodeToBytes(@NotNull byte[] src) {
        return Base64.getUrlEncoder().encode(src);
    }

    @NotNull
    public static byte[] urlEncodeToBytes(@NotNull String text) {
        return Base64.getUrlEncoder().encode(text.getBytes(Charset.forName("UTF-8")));
    }

    @NotNull
    public static byte[] urlEncodeToBytes(@NotNull ByteBuffer byteBuffer) {
        return Base64.getUrlEncoder().encode(byteBuffer.array());
    }


    public static int urlEncodeTo(@NotNull byte[] src, @NotNull byte[] dst) {
        return Base64.getUrlEncoder().encode(src, dst);
    }

    public static int urlEncodeTo(@NotNull String text, @NotNull byte[] dst) {
        return Base64.getUrlEncoder().encode(text.getBytes(Charset.forName("UTF-8")), dst);
    }

    public static int urlEncodeTo(@NotNull ByteBuffer byteBuffer, @NotNull byte[] dst) {
        return Base64.getUrlEncoder().encode(byteBuffer.array(), dst);
    }


    @NotNull
    public static ByteBuffer urlEncodeToBuffer(@NotNull byte[] src) {
        return Base64.getUrlEncoder().encode(ByteBuffer.wrap(src));
    }

    @NotNull
    public static ByteBuffer urlEncodeToBuffer(@NotNull String text) {
        return Base64.getUrlEncoder().encode(ByteBuffer.wrap(text.getBytes(Charset.forName("UTF-8"))));
    }

    @NotNull
    public static ByteBuffer urlEncodeToBuffer(@NotNull ByteBuffer byteBuffer) {
        return Base64.getUrlEncoder().encode(byteBuffer);
    }


    @NotNull
    public static String mimeEncodeToString(@NotNull byte[] src) {
        return Base64.getMimeEncoder().encodeToString(src);
    }

    @NotNull
    public static String mimeEncodeToString(@NotNull String text) {
        return Base64.getMimeEncoder().encodeToString(text.getBytes(Charset.forName("UTF-8")));
    }

    @NotNull
    public static String mimeEncodeToString(@NotNull ByteBuffer byteBuffer) {
        return Base64.getMimeEncoder().encodeToString(byteBuffer.array());
    }


    @NotNull
    public static byte[] mimeEncodeToBytes(@NotNull byte[] src) {
        return Base64.getMimeEncoder().encode(src);
    }

    @NotNull
    public static byte[] mimeEncodeToBytes(@NotNull String text) {
        return Base64.getMimeEncoder().encode(text.getBytes(Charset.forName("UTF-8")));
    }

    @NotNull
    public static byte[] mimeEncodeToBytes(@NotNull ByteBuffer byteBuffer) {
        return Base64.getMimeEncoder().encode(byteBuffer.array());
    }


    public static int mimeEncodeTo(@NotNull byte[] src, @NotNull byte[] dst) {
        return Base64.getMimeEncoder().encode(src, dst);
    }

    public static int mimeEncodeTo(@NotNull String text, @NotNull byte[] dst) {
        return Base64.getMimeEncoder().encode(text.getBytes(Charset.forName("UTF-8")), dst);
    }

    public static int mimeEncodeTo(@NotNull ByteBuffer byteBuffer, @NotNull byte[] dst) {
        return Base64.getMimeEncoder().encode(byteBuffer.array(), dst);
    }


    @NotNull
    public static ByteBuffer mimeEncodeToBuffer(@NotNull byte[] src) {
        return Base64.getMimeEncoder().encode(ByteBuffer.wrap(src));
    }

    @NotNull
    public static ByteBuffer mimeEncodeToBuffer(@NotNull String text) {
        return Base64.getMimeEncoder().encode(ByteBuffer.wrap(text.getBytes(Charset.forName("UTF-8"))));
    }

    @NotNull
    public static ByteBuffer mimeEncodeToBuffer(@NotNull ByteBuffer byteBuffer) {
        return Base64.getMimeEncoder().encode(byteBuffer);
    }


    @NotNull
    public static String mimeEncodeToString(int lineLength, @NotNull byte[] lineSeparator, @NotNull byte[] src) {
        return Base64.getMimeEncoder(lineLength, lineSeparator).encodeToString(src);
    }

    @NotNull
    public static String mimeEncodeToString(int lineLength, @NotNull byte[] lineSeparator, @NotNull String text) {
        return Base64.getMimeEncoder(lineLength, lineSeparator).encodeToString(text.getBytes(Charset.forName("UTF-8")));
    }

    @NotNull
    public static String mimeEncodeToString(int lineLength, @NotNull byte[] lineSeparator, @NotNull ByteBuffer byteBuffer) {
        return Base64.getMimeEncoder(lineLength, lineSeparator).encodeToString(byteBuffer.array());
    }


    @NotNull
    public static byte[] mimeEncodeToBytes(int lineLength, @NotNull byte[] lineSeparator, @NotNull byte[] src) {
        return Base64.getMimeEncoder(lineLength, lineSeparator).encode(src);
    }

    @NotNull
    public static byte[] mimeEncodeToBytes(int lineLength, @NotNull byte[] lineSeparator, @NotNull String text) {
        return Base64.getMimeEncoder(lineLength, lineSeparator).encode(text.getBytes(Charset.forName("UTF-8")));
    }

    @NotNull
    public static byte[] mimeEncodeToBytes(int lineLength, @NotNull byte[] lineSeparator, @NotNull ByteBuffer byteBuffer) {
        return Base64.getMimeEncoder(lineLength, lineSeparator).encode(byteBuffer.array());
    }


    public static int mimeEncodeTo(int lineLength, @NotNull byte[] lineSeparator, @NotNull byte[] src, @NotNull byte[] dst) {
        return Base64.getMimeEncoder(lineLength, lineSeparator).encode(src, dst);
    }

    public static int mimeEncodeTo(int lineLength, @NotNull byte[] lineSeparator, @NotNull String text, @NotNull byte[] dst) {
        return Base64.getMimeEncoder(lineLength, lineSeparator).encode(text.getBytes(Charset.forName("UTF-8")), dst);
    }

    public static int mimeEncodeTo(int lineLength, @NotNull byte[] lineSeparator, @NotNull ByteBuffer byteBuffer, @NotNull byte[] dst) {
        return Base64.getMimeEncoder(lineLength, lineSeparator).encode(byteBuffer.array(), dst);
    }


    @NotNull
    public static ByteBuffer mimeEncodeToBuffer(int lineLength, @NotNull byte[] lineSeparator, @NotNull byte[] src) {
        return Base64.getMimeEncoder(lineLength, lineSeparator).encode(ByteBuffer.wrap(src));
    }

    @NotNull
    public static ByteBuffer mimeEncodeToBuffer(int lineLength, @NotNull byte[] lineSeparator, @NotNull String text) {
        return Base64.getMimeEncoder(lineLength, lineSeparator).encode(ByteBuffer.wrap(text.getBytes(Charset.forName("UTF-8"))));
    }

    @NotNull
    public static ByteBuffer mimeEncodeToBuffer(int lineLength, @NotNull byte[] lineSeparator, @NotNull ByteBuffer byteBuffer) {
        return Base64.getMimeEncoder(lineLength, lineSeparator).encode(byteBuffer);
    }


    @NotNull
    public static byte[] decodeToBytes(@NotNull byte[] src) {
        return Base64.getDecoder().decode(src);
    }

    @NotNull
    public static byte[] decodeToBytes(@NotNull String text) {
        return Base64.getDecoder().decode(text.getBytes(Charset.forName("UTF-8")));
    }

    @NotNull
    public static byte[] decodeToBytes(@NotNull ByteBuffer byteBuffer) {
        return Base64.getDecoder().decode(byteBuffer).array();
    }


    @NotNull
    public static String decodeToString(@NotNull byte[] src) {
        return new String(Base64.getDecoder().decode(src), Charset.forName("UTF-8"));
    }

    @NotNull
    public static String decodeToString(@NotNull String text) {
        return new String(Base64.getDecoder().decode(text.getBytes(Charset.forName("UTF-8"))), Charset.forName("UTF-8"));
    }

    @NotNull
    public static String decodeToString(@NotNull ByteBuffer byteBuffer) {
        return new String(Base64.getDecoder().decode(byteBuffer).array(), Charset.forName("UTF-8"));
    }


    public static int decodeTo(@NotNull byte[] src, @NotNull byte[] dst) {
        return Base64.getDecoder().decode(src, dst);
    }

    public static int decodeTo(@NotNull String text, @NotNull byte[] dst) {
        return Base64.getDecoder().decode(text.getBytes(Charset.forName("UTF-8")), dst);
    }

    public static int decodeTo(@NotNull ByteBuffer byteBuffer, @NotNull byte[] dst) {
        return Base64.getDecoder().decode(byteBuffer.array(), dst);
    }


    @NotNull
    public static ByteBuffer decodeToBuffer(@NotNull byte[] src) {
        return Base64.getDecoder().decode(ByteBuffer.wrap(src));
    }

    @NotNull
    public static ByteBuffer decodeToBuffer(@NotNull String text) {
        return Base64.getDecoder().decode(ByteBuffer.wrap(text.getBytes(Charset.forName("UTF-8"))));
    }

    @NotNull
    public static ByteBuffer decodeToBuffer(@NotNull ByteBuffer byteBuffer) {
        return Base64.getDecoder().decode(byteBuffer);
    }


    @NotNull
    public static byte[] urlDecodeToBytes(@NotNull byte[] src) {
        return Base64.getUrlDecoder().decode(src);
    }

    @NotNull
    public static byte[] urlDecodeToBytes(@NotNull String text) {
        return Base64.getUrlDecoder().decode(text.getBytes(Charset.forName("UTF-8")));
    }

    @NotNull
    public static byte[] urlDecodeToBytes(@NotNull ByteBuffer byteBuffer) {
        return Base64.getUrlDecoder().decode(byteBuffer).array();
    }


    @NotNull
    public static String urlDecodeToString(@NotNull byte[] src) {
        return new String(Base64.getUrlDecoder().decode(src), Charset.forName("UTF-8"));
    }

    @NotNull
    public static String urlDecodeToString(@NotNull String text) {
        return new String(Base64.getUrlDecoder().decode(text.getBytes(Charset.forName("UTF-8"))), Charset.forName("UTF-8"));
    }

    @NotNull
    public static String urlDecodeToString(@NotNull ByteBuffer byteBuffer) {
        return new String(Base64.getUrlDecoder().decode(byteBuffer).array(), Charset.forName("UTF-8"));
    }


    public static int urlDecodeTo(@NotNull byte[] src, @NotNull byte[] dst) {
        return Base64.getUrlDecoder().decode(src, dst);
    }

    public static int urlDecodeTo(@NotNull String text, @NotNull byte[] dst) {
        return Base64.getUrlDecoder().decode(text.getBytes(Charset.forName("UTF-8")), dst);
    }

    public static int urlDecodeTo(@NotNull ByteBuffer byteBuffer, @NotNull byte[] dst) {
        return Base64.getUrlDecoder().decode(byteBuffer.array(), dst);
    }


    @NotNull
    public static ByteBuffer urlDecodeToBuffer(@NotNull byte[] src) {
        return Base64.getUrlDecoder().decode(ByteBuffer.wrap(src));
    }

    @NotNull
    public static ByteBuffer urlDecodeToBuffer(@NotNull String text) {
        return Base64.getUrlDecoder().decode(ByteBuffer.wrap(text.getBytes(Charset.forName("UTF-8"))));
    }

    @NotNull
    public static ByteBuffer urlDecodeToBuffer(@NotNull ByteBuffer byteBuffer) {
        return Base64.getUrlDecoder().decode(byteBuffer);
    }


    @NotNull
    public static byte[] mimeDecodeToBytes(@NotNull byte[] src) {
        return Base64.getMimeDecoder().decode(src);
    }

    @NotNull
    public static byte[] mimeDecodeToBytes(@NotNull String text) {
        return Base64.getMimeDecoder().decode(text.getBytes(Charset.forName("UTF-8")));
    }

    @NotNull
    public static byte[] mimeDecodeToBytes(@NotNull ByteBuffer byteBuffer) {
        return Base64.getMimeDecoder().decode(byteBuffer).array();
    }


    @NotNull
    public static String mimeDecodeToString(@NotNull byte[] src) {
        return new String(Base64.getMimeDecoder().decode(src), Charset.forName("UTF-8"));
    }

    @NotNull
    public static String mimeDecodeToString(@NotNull String text) {
        return new String(Base64.getMimeDecoder().decode(text.getBytes(Charset.forName("UTF-8"))), Charset.forName("UTF-8"));
    }

    @NotNull
    public static String mimeDecodeToString(@NotNull ByteBuffer byteBuffer) {
        return new String(Base64.getMimeDecoder().decode(byteBuffer).array(), Charset.forName("UTF-8"));
    }


    public static int mimeDecodeTo(@NotNull byte[] src, @NotNull byte[] dst) {
        return Base64.getMimeDecoder().decode(src, dst);
    }

    public static int mimeDecodeTo(@NotNull String text, @NotNull byte[] dst) {
        return Base64.getMimeDecoder().decode(text.getBytes(Charset.forName("UTF-8")), dst);
    }

    public static int mimeDecodeTo(@NotNull ByteBuffer byteBuffer, @NotNull byte[] dst) {
        return Base64.getMimeDecoder().decode(byteBuffer.array(), dst);
    }


    @NotNull
    public static ByteBuffer mimeDecodeToBuffer(@NotNull byte[] src) {
        return Base64.getMimeDecoder().decode(ByteBuffer.wrap(src));
    }

    @NotNull
    public static ByteBuffer mimeDecodeToBuffer(@NotNull String text) {
        return Base64.getMimeDecoder().decode(ByteBuffer.wrap(text.getBytes(Charset.forName("UTF-8"))));
    }

    @NotNull
    public static ByteBuffer mimeDecodeToBuffer(@NotNull ByteBuffer byteBuffer) {
        return Base64.getMimeDecoder().decode(byteBuffer);
    }
}

File: javax-kt/src/main/java/me/panpf/javaxkt/util/Base64x.kt
/*
 * Copyright (C) 2018 Peng fei Pan <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:Suppress("NOTHING_TO_INLINE")

package me.panpf.javaxkt.util

import me.panpf.javax.util.Base64x
import java.nio.ByteBuffer


/*
 * Base64 related extension methods or properties
 */


inline fun ByteArray.base64EncodeToString(): String = Base64x.encodeToString(this)

inline fun String.base64EncodeToString(): String = Base64x.encodeToString(this)

inline fun ByteBuffer.base64EncodeToString(): String = Base64x.encodeToString(this)


inline fun ByteArray.base64EncodeToBytes(): ByteArray = Base64x.encodeToBytes(this)

inline fun String.base64EncodeToBytes(): ByteArray = Base64x.encodeToBytes(this)

inline fun ByteBuffer.base64EncodeToBytes(): ByteArray = Base64x.encodeToBytes(this)


inline fun ByteArray.base64EncodeTo(dst: ByteArray): Int = Base64x.encodeTo(this, dst)

inline fun String.base64EncodeTo(dst: ByteArray): Int = Base64x.encodeTo(this, dst)

inline fun ByteBuffer.base64EncodeTo(dst: ByteArray): Int = Base64x.encodeTo(this, dst)


inline fun ByteArray.base64EncodeToBuffer(): ByteBuffer = Base64x.encodeToBuffer(this)

inline fun String.base64EncodeToBuffer(): ByteBuffer = Base64x.encodeToBuffer(this)

inline fun ByteBuffer.base64EncodeToBuffer(): ByteBuffer = Base64x.encodeToBuffer(this)


inline fun ByteArray.base64UrlEncodeToString(): String = Base64x.urlEncodeToString(this)

inline fun String.base64UrlEncodeToString(): String = Base64x.urlEncodeToString(this)

inline fun ByteBuffer.base64UrlEncodeToString(): String = Base64x.urlEncodeToString(this)


inline fun ByteArray.base64UrlEncodeToBytes(): ByteArray = Base64x.urlEncodeToBytes(this)

inline fun String.base64UrlEncodeToBytes(): ByteArray = Base64x.urlEncodeToBytes(this)

inline fun ByteBuffer.base64UrlEncodeToBytes(): ByteArray = Base64x.urlEncodeToBytes(this)


inline fun ByteArray.base64UrlEncodeTo(dst: ByteArray): Int = Base64x.urlEncodeTo(this, dst)

inline fun String.base64UrlEncodeTo(dst: ByteArray): Int = Base64x.urlEncodeTo(this, dst)

inline fun ByteBuffer.base64UrlEncodeTo(dst: ByteArray): Int = Base64x.urlEncodeTo(this, dst)


inline fun ByteArray.base64UrlEncodeToBuffer(): ByteBuffer = Base64x.urlEncodeToBuffer(this)

inline fun String.base64UrlEncodeToBuffer(): ByteBuffer = Base64x.urlEncodeToBuffer(this)

inline fun ByteBuffer.base64UrlEncodeToBuffer(): ByteBuffer = Base64x.urlEncodeToBuffer(this)


inline fun ByteArray.base64MimeEncodeToString(): String = Base64x.mimeEncodeToString(this)

inline fun String.base64MimeEncodeToString(): String = Base64x.mimeEncodeToString(this)

inline fun ByteBuffer.base64MimeEncodeToString(): String = Base64x.mimeEncodeToString(this)


inline fun ByteArray.base64MimeEncodeToBytes(): ByteArray = Base64x.mimeEncodeToBytes(this)

inline fun String.base64MimeEncodeToBytes(): ByteArray = Base64x.mimeEncodeToBytes(this)

inline fun ByteBuffer.base64MimeEncodeToBytes(): ByteArray = Base64x.mimeEncodeToBytes(this)


inline fun ByteArray.base64MimeEncodeTo(dst: ByteArray): Int = Base64x.mimeEncodeTo(this, dst)

inline fun String.base64MimeEncodeTo(dst: ByteArray): Int = Base64x.mimeEncodeTo(this, dst)

inline fun ByteBuffer.base64MimeEncodeTo(dst: ByteArray): Int = Base64x.mimeEncodeTo(this, dst)


inline fun ByteArray.base64MimeEncodeToBuffer(): ByteBuffer = Base64x.mimeEncodeToBuffer(this)

inline fun String.base64MimeEncodeToBuffer(): ByteBuffer = Base64x.mimeEncodeToBuffer(this)

inline fun ByteBuffer.base64MimeEncodeToBuffer(): ByteBuffer = Base64x.mimeEncodeToBuffer(this)


inline fun ByteArray.base64MimeEncodeToString(lineLength: Int, lineSeparator: ByteArray): String = Base64x.mimeEncodeToString(lineLength, lineSeparator, this)

inline fun String.base64MimeEncodeToString(lineLength: Int, lineSeparator: ByteArray): String = Base64x.mimeEncodeToString(lineLength, lineSeparator, this)

inline fun ByteBuffer.base64MimeEncodeToString(lineLength: Int, lineSeparator: ByteArray): String = Base64x.mimeEncodeToString(lineLength, lineSeparator, this)


inline fun ByteArray.base64MimeEncodeToBytes(lineLength: Int, lineSeparator: ByteArray): ByteArray = Base64x.mimeEncodeToBytes(lineLength, lineSeparator, this)

inline fun String.base64MimeEncodeToBytes(lineLength: Int, lineSeparator: ByteArray): ByteArray = Base64x.mimeEncodeToBytes(lineLength, lineSeparator, this)

inline fun ByteBuffer.base64MimeEncodeToBytes(lineLength: Int, lineSeparator: ByteArray): ByteArray = Base64x.mimeEncodeToBytes(lineLength, lineSeparator, this)


inline fun ByteArray.base64MimeEncodeTo(lineLength: Int, lineSeparator: ByteArray, dst: ByteArray): Int = Base64x.mimeEncodeTo(lineLength, lineSeparator, this, dst)

inline fun String.base64MimeEncodeTo(lineLength: Int, lineSeparator: ByteArray, dst: ByteArray): Int = Base64x.mimeEncodeTo(lineLength, lineSeparator, this, dst)

inline fun ByteBuffer.base64MimeEncodeTo(lineLength: Int, lineSeparator: ByteArray, dst: ByteArray): Int = Base64x.mimeEncodeTo(lineLength, lineSeparator, this, dst)


inline fun ByteArray.base64MimeEncodeToBuffer(lineLength: Int, lineSeparator: ByteArray): ByteBuffer = Base64x.mimeEncodeToBuffer(lineLength, lineSeparator, this)

inline fun String.base64MimeEncodeToBuffer(lineLength: Int, lineSeparator: ByteArray): ByteBuffer = Base64x.mimeEncodeToBuffer(lineLength, lineSeparator, this)

inline fun ByteBuffer.base64MimeEncodeToBuffer(lineLength: Int, lineSeparator: ByteArray): ByteBuffer = Base64x.mimeEncodeToBuffer(lineLength, lineSeparator, this)


inline fun ByteArray.base64DecodeToBytes(): ByteArray = Base64x.decodeToBytes(this)

inline fun String.base64DecodeToBytes(): ByteArray = Base64x.decodeToBytes(this)

inline fun ByteBuffer.base64DecodeToBytes(): ByteArray = Base64x.decodeToBytes(this)


inline fun ByteArray.base64DecodeToString(): String = Base64x.decodeToString(this)

inline fun String.base64DecodeToString(): String = Base64x.decodeToString(this)

inline fun ByteBuffer.base64DecodeToString(): String = Base64x.decodeToString(this)


inline fun ByteArray.base64DecodeTo(dst: ByteArray): Int = Base64x.decodeTo(this, dst)

inline fun String.base64DecodeTo(dst: ByteArray): Int = Base64x.decodeTo(this, dst)

inline fun ByteBuffer.base64DecodeTo(dst: ByteArray): Int = Base64x.decodeTo(this, dst)


inline fun ByteArray.base64DecodeToBuffer(): ByteBuffer = Base64x.decodeToBuffer(this)

inline fun String.base64DecodeToBuffer(): ByteBuffer = Base64x.decodeToBuffer(this)

inline fun ByteBuffer.base64DecodeToBuffer(): ByteBuffer = Base64x.decodeToBuffer(this)


inline fun ByteArray.base64UrlDecodeToBytes(): ByteArray = Base64x.urlDecodeToBytes(this)

inline fun String.base64UrlDecodeToBytes(): ByteArray = Base64x.urlDecodeToBytes(this)

inline fun ByteBuffer.base64UrlDecodeToBytes(): ByteArray = Base64x.urlDecodeToBytes(this)


inline fun ByteArray.base64UrlDecodeToString(): String = Base64x.urlDecodeToString(this)

inline fun String.base64UrlDecodeToString(): String = Base64x.urlDecodeToString(this)

inline fun ByteBuffer.base64UrlDecodeToString(): String = Base64x.urlDecodeToString(this)


inline fun ByteArray.base64UrlDecodeTo(dst: ByteArray): Int = Base64x.urlDecodeTo(this, dst)

inline fun String.base64UrlDecodeTo(dst: ByteArray): Int = Base64x.urlDecodeTo(this, dst)

inline fun ByteBuffer.base64UrlDecodeTo(dst: ByteArray): Int = Base64x.urlDecodeTo(this, dst)


inline fun ByteArray.base64UrlDecodeToBuffer(): ByteBuffer = Base64x.urlDecodeToBuffer(this)

inline fun String.base64UrlDecodeToBuffer(): ByteBuffer = Base64x.urlDecodeToBuffer(this)

inline fun ByteBuffer.base64UrlDecodeToBuffer(): ByteBuffer = Base64x.urlDecodeToBuffer(this)


inline fun ByteArray.base64MimeDecodeToBytes(): ByteArray = Base64x.mimeDecodeToBytes(this)

inline fun String.base64MimeDecodeToBytes(): ByteArray = Base64x.mimeDecodeToBytes(this)

inline fun ByteBuffer.base64MimeDecodeToBytes(): ByteArray = Base64x.mimeDecodeToBytes(this)


inline fun ByteArray.base64MimeDecodeToString(): String = Base64x.mimeDecodeToString(this)

inline fun String.base64MimeDecodeToString(): String = Base64x.mimeDecodeToString(this)

inline fun ByteBuffer.base64MimeDecodeToString(): String = Base64x.mimeDecodeToString(this)


inline fun ByteArray.base64MimeDecodeTo(dst: ByteArray): Int = Base64x.mimeDecodeTo(this, dst)

inline fun String.base64MimeDecodeTo(dst: ByteArray): Int = Base64x.mimeDecodeTo(this, dst)

inline fun ByteBuffer.base64MimeDecodeTo(dst: ByteArray): Int = Base64x.mimeDecodeTo(this, dst)


inline fun ByteArray.base64MimeDecodeToBuffer(): ByteBuffer = Base64x.mimeDecodeToBuffer(this)

inline fun String.base64MimeDecodeToBuffer(): ByteBuffer = Base64x.mimeDecodeToBuffer(this)

inline fun ByteBuffer.base64MimeDecodeToBuffer(): ByteBuffer = Base64x.mimeDecodeToBuffer(this)
